### PR TITLE
Add EndpointMetadataCollection.GetRequiredMetadata #39921

### LINKED ISSUE #39921: Add EndpointMetadataCollection.GetRequiredMetadata
## Background and Motivation

`EndpointMetadataCollection` has a `GetMetadata<T>()` method:

```csharp
public T? GetMetadata<T> () where T : class;
```

https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.endpointmetadatacollection.getmetadata?view=aspnetcore-6.0#microsoft-aspnetcore-http-endpointmetadatacollection-getmetadata-1

It returns null if there is no matching metadata. It would be useful to have a `GetRequiredMetadata<T>()` method that returns a value or throws.

## Proposed API

```diff
namespace Microsoft.AspNetCore.Http
{
    public class EndpointMetadataCollection
    {
+       public T GetRequiredMetadata<T> () where T : class;
    }
}
```

The method would function the same as getting a required service from DI or a required feature from a feature collection.

## Usage Examples

```csharp
endpoints.MapGet("/", context =>
{
    var endpoint = context.GetEndpoint();
    var requiredMetadata = endpoint.Metadata.GetRequiredMetadata<MyMetadata>();
});
```

### PR DESCRIPTION
**Pull request by paw-sad**
# Add EndpointMetadataCollection.GetRequiredMetadata

I've added GetRequiredMetadata<T>() method to  Microsoft.AspNetCore.Http.EndpointMetadataCollection. It throws InvalidOperationException when metadata of type T is not found.

## Description
Fixes #39921

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=40321) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=40321) before we can accept your contribution.<br/><br/>:x: paw-sad [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=40321)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=40321) it.</sub>

**Comment by paw-sad** [comment]
Thanks a lot for fix @pranavkm ! I couldn't figure it out! Is there some documentation for this? How did you know about the 'forwarded, contained in'?

**Comment by pranavkm** [comment]
@paw-sad I'd previously touched the file so I happened to know what to do with the analyzer.

File: src/Http/Http.Abstractions/src/Routing/EndpointMetadataCollection.cs [hunk 1]
```diff
@@ -126,6 +126,20 @@ public EndpointMetadataCollection(params object[] items)
...
 
+    /// <summary>
+    /// Gets the most significant metadata item of type <typeparamref name="T"/>.
+    /// Throws an <see cref="InvalidOperationException"/> if the metadata is not found.
+    /// </summary>
+    /// <typeparam name="T">The type of metadata to retrieve.</typeparam>
+    /// <returns>
+    /// The most significant metadata of type <typeparamref name="T"/>.
+    /// </returns>
+    public T GetRequiredMetadata<T>() where T : class
```
Comment (Kahbazi) [inline src/Http/Http.Abstractions/src/Routing/EndpointMetadataCollection.cs:137]:
```suggestion
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public T GetRequiredMetadata<T>() where T : class
```
Does this method also need `AggressiveInlining` like the other one?
https://github.com/dotnet/aspnetcore/blob/0bc20453b290fa72a913fe4744cdcdff64663a1e/src/Http/Http.Abstractions/src/Routing/EndpointMetadataCollection.cs#L72-L73

Comment (pranavkm) [reply src/Http/Http.Abstractions/src/Routing/EndpointMetadataCollection.cs:137]: AFAIK throwing an exception prevents a method from being inlined.

**Review by pranavkm — APPROVED** [review]

**Comment by pranavkm** [comment]
Thanks!

**Pull request merged**